# Pheme training

Original Pheme repo that supports training Ukrainian
https://github.com/taras-sereda/pheme/tree/train-ua

T2S
https://github.com/taras-sereda/pheme/tree/train-ua

S2A
https://github.com/taras-sereda/pheme/blob/train-ua/train_s2a.py

In [1]:
!git clone --branch train-ua https://github.com/taras-sereda/pheme.git

Cloning into 'pheme'...
remote: Enumerating objects: 230, done.
remote: Counting objects: 100% (58/58), done.
remote: Compressing objects: 100% (36/36), done.
remote: Total 230 (delta 35), reused 35 (delta 22), pack-reused 172
Receiving objects: 100% (230/230), 10.04 MiB | 8.50 MiB/s, done.
Resolving deltas: 100% (69/69), done.


In [2]:
!pip3 install torch torchvision torchaudio
!pip3 install docker_pycreds
!pip3 install sentry_sdk
!pip3 install -U wandb>=0.12.10

  Using cached nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (23.7 MB)
  Using cached nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (823 kB)
  Using cached nvidia_cuda_cupti_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (14.1 MB)
  Using cached nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl (731.7 MB)
  Using cached nvidia_cublas_cu12-12.1.3.1-py3-none-manylinux1_x86_64.whl (410.6 MB)
  Using cached nvidia_cufft_cu12-11.0.2.54-py3-none-manylinux1_x86_64.whl (121.6 MB)
  Using cached nvidia_curand_cu12-10.3.2.106-py3-none-manylinux1_x86_64.whl (56.5 MB)
  Using cached nvidia_cusolver_cu12-11.4.5.107-py3-none-manylinux1_x86_64.whl (124.2 MB)
  Using cached nvidia_cusparse_cu12-12.1.0.106-py3-none-manylinux1_x86_64.whl (196.0 MB)
  Using cached nvidia_nccl_cu12-2.19.3-py3-none-manylinux1_x86_64.whl (166.0 MB)
  Using cached nvidia_nvtx_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (99 kB)
  Using cached nvidia_nvjitlink_cu12-12.4.127-py3-none-m

 You have to install espeak 1.51. It's important since starting from this version the package supports Ukrainain

In [3]:
!git clone https://github.com/espeak-ng/espeak-ng.git

Cloning into 'espeak-ng'...
remote: Enumerating objects: 48498, done.
remote: Counting objects: 100% (502/502), done.
remote: Compressing objects: 100% (309/309), done.
remote: Total 48498 (delta 245), reused 367 (delta 172), pack-reused 47996
Receiving objects: 100% (48498/48498), 59.76 MiB | 13.59 MiB/s, done.
Resolving deltas: 100% (33541/33541), done.


In [4]:
!apt-get install make autoconf automake libtool pkg-config
!apt-get install gcc
!apt-get install libsonic-dev
!apt-get install ronn
!apt-get install kramdown

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
autoconf is already the newest version (2.71-2).
autoconf set to manually installed.
automake is already the newest version (1:1.16.5-1.3).
automake set to manually installed.
make is already the newest version (4.3-4.1build1).
make set to manually installed.
pkg-config is already the newest version (0.29.2-1ubuntu3).
Suggested packages:
  libtool-doc gcj-jdk
The following NEW packages will be installed:
  libtool
0 upgraded, 1 newly installed, 0 to remove and 45 not upgraded.
Need to get 164 kB of archives.
After this operation, 1,227 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu jammy/main amd64 libtool all 2.4.6-15build2 [164 kB]
Fetched 164 kB in 1s (113 kB/s)
Selecting previously unselected package libtool.
(Reading database ... 121920 files and directories currently installed.)
Preparing to unpack .../libtool_2.4.6-15build2_all.deb ...
Unpacking libt

In [5]:
%cd espeak-ng

/content/espeak-ng


In [6]:
!./autogen.sh
!./configure --prefix=/usr

libtoolize: putting auxiliary files in '.'.
libtoolize: copying file './ltmain.sh'
libtoolize: putting macros in AC_CONFIG_MACRO_DIRS, 'm4'.
libtoolize: copying file 'm4/libtool.m4'
libtoolize: copying file 'm4/ltoptions.m4'
libtoolize: copying file 'm4/ltsugar.m4'
libtoolize: copying file 'm4/ltversion.m4'
libtoolize: copying file 'm4/lt~obsolete.m4'
configure.ac:4: installing './compile'
configure.ac:4: installing './config.guess'
configure.ac:4: installing './config.sub'
configure.ac:3: installing './install-sh'
configure.ac:3: installing './missing'
Makefile.am:112: warning: shell dirname $@: non-POSIX variable name
Makefile.am:112: (probably a GNU make extension)
Makefile.am:115: warning: shell dirname $@: non-POSIX variable name
Makefile.am:115: (probably a GNU make extension)
Makefile.am:117: warning: wildcard docs/*.md: non-POSIX variable name
Makefile.am:117: (probably a GNU make extension)
Makefile.am:117: warning: wildcard docs/*/*.md: non-POSIX variable name
Makefile.am:117

In [7]:
!make

Makefile:3123: warning: ignoring prerequisites on suffix rule definition
make  all-am
make[1]: Entering directory '/content/espeak-ng'
Makefile:3123: warning: ignoring prerequisites on suffix rule definition
  CC       src/speak-ng.o
In file included from src/speak-ng.c:23:
src/espeak-ng.c: In function ‘main’:
src/espeak-ng.c:777:17: warning: ignoring return value of ‘fread’ declared with attribute ‘warn_unused_result’ []8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-result-Wunused-result]8;;]
  777 |                 fread(p_text, 1, filesize, f_text);
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  CC       src/ucd-tools/src/libespeak_ng_la-case.lo
  CC       src/ucd-tools/src/libespeak_ng_la-categories.lo
  CC       src/ucd-tools/src/libespeak_ng_la-ctype.lo
  CC       src/ucd-tools/src/libespeak_ng_la-proplist.lo
  CC       src/ucd-tools/src/libespeak_ng_la-scripts.lo
  CC       src/ucd-tools/src/libespeak_ng_la-tostring.lo
  CC       src/l

In [8]:
!ESPEAK_DATA_PATH=`pwd` LD_LIBRARY_PATH=src:${LD_LIBRARY_PATH} src/espeak-ng ...
!make LIBDIR=/usr/lib/x86_64-linux-gnu install

Makefile:3123: warning: ignoring prerequisites on suffix rule definition
make: Circular phsource/phonemes.stamp <- phsource/phonemes.stamp dependency dropped.
make[1]: Entering directory '/content/espeak-ng'
Makefile:3123: warning: ignoring prerequisites on suffix rule definition
 /usr/bin/mkdir -p '/usr/lib'
 /bin/bash ./libtool   --mode=install /usr/bin/install -c   src/libespeak-ng.la '/usr/lib'
libtool: install: /usr/bin/install -c src/.libs/libespeak-ng.so.1.1.51 /usr/lib/libespeak-ng.so.1.1.51
libtool: install: (cd /usr/lib && { ln -s -f libespeak-ng.so.1.1.51 libespeak-ng.so.1 || { rm -f libespeak-ng.so.1 && ln -s libespeak-ng.so.1.1.51 libespeak-ng.so.1; }; })
libtool: install: (cd /usr/lib && { ln -s -f libespeak-ng.so.1.1.51 libespeak-ng.so || { rm -f libespeak-ng.so && ln -s libespeak-ng.so.1.1.51 libespeak-ng.so; }; })
libtool: install: /usr/bin/install -c src/.libs/libespeak-ng.lai /usr/lib/libespeak-ng.la
libtool: install: /usr/bin/install -c src/.libs/libespeak-ng.a /usr

In [9]:
!find /usr/lib | grep libespeak-ng
!espeak --version

/usr/lib/libespeak-ng.so
/usr/lib/libespeak-ng.so.1
/usr/lib/libespeak-ng.so.1.1.51
/usr/lib/libespeak-ng.a
/usr/lib/libespeak-ng.la
eSpeak NG text-to-speech: 1.52-dev  Data at: /usr/share/espeak-ng-data


You suppose to see: eSpeak NG text-to-speech: 1.52-dev  Data at: /usr/share/espeak-ng-data

In [10]:
%cd ../pheme/

/content/pheme


In [11]:
!pip3 install -r requirements.txt --no-deps

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.5/14.5 MB 7.3 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 24.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 230.6/230.6 kB 30.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 996.9/996.9 kB 74.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 117.0/117.0 kB 18.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.2/61.2 kB 9.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.1/10.1 MB 103.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 162.5/162.5 kB 23.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 86.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 310.7/310.7 kB 37.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.7/56.7 kB 8.2 MB/s eta 0:00:00
  Preparing metadata (s

In [ ]:
# you have to configure hugging face token in order to use pyannote
# if you don't have it - https://huggingface.co/settings/tokens

%env HUGGING_FACE_HUB_TOKEN=YOUR_HF_TOKEN

In [ ]:
%cd pheme/

!mkdir -p "ckpt/speechtokenizer"
%cd "ckpt/speechtokenizer"

In [ ]:
!wget "https://huggingface.co/fnlp/SpeechTokenizer/resolve/main/speechtokenizer_hubert_avg/SpeechTokenizer.pt"
!wget "https://huggingface.co/fnlp/SpeechTokenizer/resolve/main/speechtokenizer_hubert_avg/config.json"
%cd ..
!wget "https://huggingface.co/fnlp/USLM/resolve/main/USLM_libritts/unique_text_tokens.k2symbols"

In [ ]:
!git clone https://huggingface.co/PolyAI/pheme_small pheme

In [13]:
!mkdir -p "t2s"
!mkdir -p "s2a"
!mv pheme/config_t2s.json t2s/config.json
!mv pheme/generation_config.json t2s/generation_config.json
!mv pheme/t2s.bin t2s/pytorch_model.bin
!mv pheme/config_s2a.json s2a/config.json
!mv pheme/s2a.ckpt s2a/s2a.ckpt

In [14]:
%cd ..

/content/pheme


In [ ]:
# check if everything is set up correctly
!python transformer_infer.py

The next step is to get training data. We are going to use LADA dataset which might find here: https://github.com/egorsmkv/ukrainian-tts-datasets/tree/main/lada

It's important to name folder with audios as audios. It's hard coded in the Pheme library :)

Download 16khz version or you have to resample audios

In [18]:
PATH_TO_TRAINING_DATA = "/content/drive/MyDrive/Projects/audio-course/dataset_lada_16khz/"

In [ ]:
!PYTHONPATH=. python utils/get_tokens_speech_tokenizer.py \
    --config_path ckpt/speechtokenizer/config.json \
    --ckpt_path ckpt/speechtokenizer/SpeechTokenizer.pt \
    --encoding_input "{PATH_TO_TRAINING_DATA}audios" \
    --encoding_output "{PATH_TO_TRAINING_DATA}audios-speech-tokenizer"

# TODO (have to do to receive full credit)
## Explain what is the roll of get_tokens_speech_tokenizer?

In [ ]:
!PYTHONPATH="." python utils/data_prep.py \
--dataset lada \
--data_root "{PATH_TO_TRAINING_DATA}" \
--lang uk

# TODO (have to do to receive full credit)
## Explain what is the roll of data_prep?

# TODO: Train t2s

In [ ]:
TRAIN_MANIFEST = PATH_TO_TRAINING_DATA + "train.json"
DEV_MANIFEST = PATH_TO_TRAINING_DATA + "dev.json"
TEXT_TOKENS_PATH = PATH_TO_TRAINING_DATA + "unique_text_tokens.k2symbols"

OUT_DIR="experiments/t2s-lada"

!python train_t2s.py --metapath "{TRAIN_MANIFEST}" \
  --val_metapath "{DEV_MANIFEST}" \
  --text_tokens_file "{TEXT_TOKENS_PATH}" \
  --output_dir "{OUT_DIR}" \
  --model_size tiny --batch_size 64 \
  --nworkers 12 --warmup_steps 10000 \
  --save_steps 500 --n_epochs 100 \
  --learning_rate 1e-3

# Train s2a

In [ ]:
!python train_s2a.py --saving_path "{OUT_DIR}" --sampledir "{OUT_DIR}" --vocoder_type SPEECHTOKENIZER \
 --n_codes 1024 --n_cluster_groups 7 --metapath "{TRAIN_MANIFEST}" \
 --val_metapath "{DEV_MANIFEST}" \
 --warmup_step 10000 --nworkers 12 --first_n_lvls 7 \
 --batch_size 200 --ffd_size 1024 --hidden_size 768 --enc_nlayers 3 --dec_nlayers 6 --nheads 8 \
 --depthwise_conv_kernel_size 5 \
 --val_check_interval 60 --sample_rate 16000 --lr 5e-4 \
 --check_val_every_n_epoch 1 --n_semantic_codes 1024 \
 --distributed

# HW


Theory (5 points):

* Answer theory questions in the Google Form.

Practice (10 points):

* Explore code and answer TODOs - 2 points
* Trains t2s and s2a models - 2 points
* Explore training curves and try different hyper-parameters (at least 3 combinations). Finally, make conclusions from training plots - 3 points
* Generate audio samples and compare between different runs. Add to final report - 2 points

